### PR TITLE
Issue #3110: Modified Main to remove hardcoded messages and to support i18n for the messages

### PR DESCRIPTION
**Pull request by subkrish**
Issue #3110

The hardcoded messages in Main have been removed and i18n support for those messages have been added.

### CODE REVIEW

**Comment by subkrish** [comment]
@romani @rnveach Please review.

**Comment by romani** [comment]
@subkrish ,

thanks a lot, and that that you are planning to do this step by step
BUT before we proceed with such process, lets find a reason of 

> please investigate why DefaultLoggerTest is NOT failed in localized UTs, as "Error auditing" is hard coded in UT in English.

see mu comment at https://github.com/checkstyle/checkstyle/issues/3110#issuecomment-311208795

**Comment by subkrish** [comment]
@romani There were no errors thrown because in MainTest this option is set:

```
    @BeforeClass
    public static void init() {
        // Set locale to root to prevent check message fail
        // in other language environment.
        Locale.setDefault(Locale.ROOT);
    }
```
This will never allow it to fail when tested on another language.

**Comment by romani** [comment]
> Should this be removed so that they can be tested for localized messages?

yes.

**Comment by subkrish** [comment]
@romani @rnveach 
I'm stuck in a situation.

In this method
https://github.com/checkstyle/checkstyle/blob/master/src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java#L438-#L467

```
            assertTrue(ex.getCause().getMessage()
                    .startsWith("Unable to load properties from file '"));
            assertTrue(ex.getCause().getMessage().endsWith(":invalid'."));
```
I am unable to set the LocalizedMessage because we are checking only if the message starts with that sentence in English. How would I be able to split the LocalizedMessage to check if it was right or not? There are similar cases in other methods in the MainTest file.

**Comment by romani** [comment]
> Can I change it from assertTrue to assertEquals so that I can check the complete message or is there any other way?

yes, it is always better to compare whole string.

**Comment by subkrish** [comment]
@romani @rnveach 

How should I handle this method -> `testLoadPropertiesIoException()`?
```
    public void testLoadPropertiesIoException() throws Exception {
        final Class<?>[] param = new Class<?>[1];
        param[0] = File.class;
        final Method method = Main.class.getDeclaredMethod("loadProperties", param);
        method.setAccessible(true);
        final LocalizedMessage loadPropertiesMessage = new LocalizedMessage(0,
                Definitions.CHECKSTYLE_BUNDLE, "Main.loadProperties",
                new String[] {"/\0:invalid"}, null, getClass(), null);
        try {
            if (System.getProperty("os.name").toLowerCase(Locale.ENGLISH).startsWith("windows")) {
                // https://support.microsoft.com/en-us/kb/177506 but this only for NTFS
                // WindowsServer 2012 use Resilient File System (ReFS), so any name is ok
                final File file = new File(File.separator + ":invalid");
                if (file.exists()) {
                    file.delete();
                }
                method.invoke(null, new File(file.getAbsolutePath()));
            }
            else {
                method.invoke(null, new File(File.separator + "\0:invalid"));
            }
            fail("Exception was expected");
        }
        catch (InvocationTargetException ex) {
            assertTrue(ex.getCause() instanceof CheckstyleException);
            // We do separate validation for message as in Windows
            // disk drive letter appear in message,
            // so we skip that drive letter for compatibility issues
            assertEquals(loadPropertiesMessage.getMessage(), ex.getCause().getLocalizedMessage());
        }
    }
```
This is the method with the changes I have made.

Travis Ci is ok but Appveyor is failing. I'm pretty sire this has something to do with the different file structure that is present in Windows and Unix which is mentioned in the method already.

**Comment by romani** [comment]
@subkrish , please do `endsWith("invalid")`

**Comment by subkrish** [comment]
@romani Should I remove `assertEquals(loadPropertiesMessage.getMessage(), ex.getCause().getLocalizedMessage());` and then add `endsWith("invalid")`?

**Comment by subkrish** [comment]
@romani @rnveach I can't do `            assertTrue(ex.getCause().getMessage().endsWith(":invalid'."));` because it is not necessary that in a different it language that it may come at the end. 

Like in German the message is `Die Eigenschaften von Datei ':invalid' können nicht geladen werden.` So here it does not end with ':invalid', infact it comes in the center. Is there any other that is possible?

**Comment by romani** [comment]
please rebase on latest master. 
after each release it is mandatory to rebase to avoid CIs failures.

==========

lets try to make it as `contains("invalid")`

**Comment by subkrish** [comment]
@romani Rebased. Done. Please review.

**Comment by subkrish** [comment]
@romani @rnveach Rebased again. Done. Please review.

File: src/main/resources/com/puppycrawl/tools/checkstyle/messages_zh.properties [hunk 1]
```diff
@@ -8,6 +8,13 @@ DefaultLogger.addException=检查错误： {0}
 DefaultLogger.auditStarted=开始检查……
 DefaultLogger.auditFinished=检查完成。
 
+#
+# for Main.java
+#
+Main.errorCounter=Checkstyle以 {0} 個錯誤結束。
+Main.loadProperties=無法從文件 ''{0}'' 加載屬性。
+Main.createListener=輸出格式無效 找到 ''{0}''，但預期為 ''{1}'' 或 ''{2}'' 。
```
Comment (romani) [inline src/main/resources/com/puppycrawl/tools/checkstyle/messages_zh.properties]: @Luolc , can you recheck this translation and provide better working if required.

File: src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java [hunk 2]
```diff
@@ -302,14 +324,31 @@ public void testExistingTargetFileWithError()
             throws Exception {
         exit.expectSystemExitWithStatus(2);
         exit.checkAssertionAfterwards(() -> {
+            final LocalizedMessage auditStartMessage = new LocalizedMessage(0,
+                    Definitions.CHECKSTYLE_BUNDLE, "DefaultLogger.auditStarted",
+                    null, null, getClass(), null);
+            final LocalizedMessage auditFinishMessage = new LocalizedMessage(0,
+                    Definitions.CHECKSTYLE_BUNDLE, "DefaultLogger.auditFinished",
+                    null, null, getClass(), null);
```
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java]: @rnveach , I do not like class fields for Tests, but I think that moving auditStartMessage/auditFinishMessage and other string without parameters to Test classfields is reasonable.

File: src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java [hunk 3]
```diff
@@ -835,8 +911,11 @@ public void testCustomSimpleRootModule() throws Exception {
     public void testExecuteIgnoredModule() throws Exception {
         exit.expectSystemExitWithStatus(-2);
         exit.checkAssertionAfterwards(() -> {
+            final LocalizedMessage errorCounterMessage = new LocalizedMessage(0,
+                    Definitions.CHECKSTYLE_BUNDLE, "Main.errorCounter",
+                    new String[] {String.valueOf(1)}, null, getClass(), null);
```
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java]: Main.errorCounter with {1} is used about 7 times, I propose to make a field for it too, as very frequent element.

File: src/main/resources/com/puppycrawl/tools/checkstyle/messages_zh.properties (hunk 1 above)
Comment (Luolc) [inline src/main/resources/com/puppycrawl/tools/checkstyle/messages_zh.properties]: Almost good, expect that [Simplified Chinese](https://en.wikipedia.org/wiki/Simplified_Chinese_characters) is more common. So my suggestion:
```
Main.errorCounter=Checkstyle以 {0} 个错误结束。
Main.loadProperties=无法从文件 ''{0}'' 中加载属性。
Main.createListener=非法的输出格式。预期为 ''{1}'' 或 ''{2}''，但实际为 ''{0}''。
```

File: src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java (hunk 2 above)
Comment (subkrish) [inline src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java]: @romani @rnveach Should I move forward with making these changes?

Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java]: @romani Like I said below at https://github.com/checkstyle/checkstyle/pull/4553#discussion_r125308406 ,
These keys should be fields like we do for checks and their tests. I am fine with this being a public field in the `Main` class. This is similar and follows the pattern on how the checks are done.

File: src/main/resources/com/puppycrawl/tools/checkstyle/messages_zh.properties (hunk 1 above)
Comment (rnveach) [inline src/main/resources/com/puppycrawl/tools/checkstyle/messages_zh.properties]: @subkrish Please us Luolc's changes.

Comment (subkrish) [inline src/main/resources/com/puppycrawl/tools/checkstyle/messages_zh.properties]: Done.

File: src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java (hunk 2 above)
Comment (subkrish) [inline src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java]: Done

File: src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java (hunk 3 above)
Comment (subkrish) [inline src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java]: Done.

**Comment by subkrish** [comment]
@romani @rnveach Changes have been made. Done. Please review.

File: src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java [hunk 4]
```diff
@@ -269,8 +277,14 @@ public void testExistingTargetFileXmlOutput() throws Exception {
...
 
         exit.checkAssertionAfterwards(() -> {
-            assertEquals(String.format(Locale.ROOT, "Starting audit...%n"
-                    + "Audit done.%n"), systemOut.getLog());
+            final LocalizedMessage auditStartMessage = new LocalizedMessage(0,
+                    Definitions.CHECKSTYLE_BUNDLE, AUDIT_STARTED_MESSAGE,
+                    null, null, getClass(), null);
+            final LocalizedMessage auditFinishMessage = new LocalizedMessage(0,
+                    Definitions.CHECKSTYLE_BUNDLE, AUDIT_FINISHED_MESSAGE,
+                    null, null, getClass(), null);
```
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java]: @subkrish , please move 
```
            final LocalizedMessage auditStartMessage = new LocalizedMessage(0,
                    Definitions.CHECKSTYLE_BUNDLE, AUDIT_STARTED_MESSAGE,
                    null, null, getClass(), null);
            final LocalizedMessage auditFinishMessage = new LocalizedMessage(0,
                    Definitions.CHECKSTYLE_BUNDLE, AUDIT_FINISHED_MESSAGE,
                    null, null, getClass(), null);
```
to be class fields, that locallized messages is not changeable, so they are like a constants.

File: src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java [hunk 5]
```diff
@@ -405,8 +459,11 @@ public void testExistingIncorrectChildrenInConfigFile()
             throws Exception {
         exit.expectSystemExitWithStatus(-2);
         exit.checkAssertionAfterwards(() -> {
+            final LocalizedMessage errorCounterMessage = new LocalizedMessage(0,
+                    Definitions.CHECKSTYLE_BUNDLE, ERROR_COUNTER_MESSAGE,
+                    new String[] {String.valueOf(1)}, null, getClass(), null);
```
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java]: please move this variable to class field and use it in all cases. You have already 7 of them

File: src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java (hunk 4 above)
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java]: @romani I advise against assigning these to a class field. These classes are based on static variable `locale` at https://github.com/checkstyle/checkstyle/blob/4a087c020c2b97130e94f9292b94cb4bc1f6ed93/src/main/java/com/puppycrawl/tools/checkstyle/api/LocalizedMessage.java#L64 .
If we are changing it's values in a test's lifecycle, having it as a field will make it's values based on the time the class is instantiated and not when the test is run. If another test is manipulating any values, we could be stuck with improper values when the class is instantiated.

Comment (subkrish) [inline src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java]: @rnveach @romani Should I leave it as is or make the change?

Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java]: @subkrish Continue on as romani has requested.

**Comment by rnveach** [comment]
@subkrish This branch has conflicts, please fix.

**Comment by subkrish** [comment]
@romani @rnveach Resolved. Please review. Done.

**Comment by subkrish** [comment]
I made an inner class 
```
    private final class MyAssertion implements Assertion {
        private final String[][] outputValues;
        private final String bundle;
        private final String msgKey;
        private final int allowedLength;

        private MyAssertion(String[][] outputValues, String bundle, String msgKey, int
                allowedLength) {
            this.outputValues = outputValues;
            this.bundle = bundle;
            this.msgKey = msgKey;
            this.allowedLength = allowedLength;
        }

        @Override
        public void checkAssertion() throws IOException {
            final String expectedPath = getFilePath("main/") + File.separator;
            final StringBuilder sb = new StringBuilder(28);
            sb.append(auditStartMessage.getMessage())
                    .append(System.getProperty("line.separator"));
            final String format = "[WARN] %s.java:%s: %s [FileLength]";
            for (String[] outputValue : outputValues) {
                final String localizedMessage = new LocalizedMessage(0, bundle,
                        msgKey, new Integer[] {Integer.valueOf(outputValue[2]), allowedLength},
                        null, getClass(), null).getMessage();
                final String line = String.format(Locale.ROOT, format,
                        expectedPath + outputValue[0], outputValue[1], localizedMessage);
                sb.append(line).append(System.getProperty("line.separator"));
            }
            sb.append(auditFinishMessage.getMessage())
                    .append(System.getProperty("line.separator"));
            assertEquals(sb.toString(), systemOut.getLog());
            assertEquals("", systemErr.getLog());
        }
    }
```

This is for `testExistingDirectoryWithViolations()`. If I use public modifier at `public MyAssertion(String[][] outputValues, String bundle, String msgKey, int
                allowedLength)`, I get this error -> `Redundant 'public' modifier. [RedundantModifier]`.

But if I use a private modifier `private MyAssertion(String[][] outputValues, String bundle, String msgKey, int allowedLength)`, I get this error -> `PMD Failure: com.puppycrawl.tools.checkstyle.MainTest:552 Rule:AccessorClassGeneration Priority:3 Avoid instantiation through private constructors from outside of the constructors class..`

What can I do in this situation?

**Comment by rnveach** [comment]
@subkrish Try removing modifier on consutructor like https://github.com/checkstyle/checkstyle/blob/e67e59bce1940ddbbd44713289dc2368e56b6d15/src/test/java/com/puppycrawl/tools/checkstyle/XMLLoggerTest.java#L260 .

**Comment by subkrish** [comment]
@romani @rnveach Looks like it worked. Done. Please review.

File: src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java (hunk 5 above)
Comment (subkrish) [inline src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java]: Done.

**Comment by subkrish** [comment]
@romani @rnveach Please review. Done.

File: src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java [hunk 6]
```diff
@@ -100,6 +98,18 @@
...
 
+    private final LocalizedMessage auditStartMessage = new LocalizedMessage(0,
+            Definitions.CHECKSTYLE_BUNDLE, "DefaultLogger.auditStarted", null, null,
+            getClass(), null);
+
+    private final LocalizedMessage auditFinishMessage = new LocalizedMessage(0,
+            Definitions.CHECKSTYLE_BUNDLE, "DefaultLogger.auditFinished", null, null,
+            getClass(), null);
+
+    private final LocalizedMessage errorCounterMessage = new LocalizedMessage(0,
```
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java]: please name it "errorCounterOneMessage" to clearly define that is not general message, but parametrized by "1"

File: src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java [hunk 7]
```diff
@@ -226,8 +229,8 @@ public void testNonExistingOutputFormat() throws Exception {
     public void testNonExistingClass() throws Exception {
         exit.expectSystemExitWithStatus(-2);
         exit.checkAssertionAfterwards(() -> {
-            final String expectedExceptionMessage =
-                    String.format(Locale.ROOT, "Checkstyle ends with 1 errors.%n");
+            final String expectedExceptionMessage = errorCounterMessage.getMessage()
+                    + System.getProperty("line.separator");
```
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java]: `System.getProperty("line.separator")` is used too much , please move it to field with name `EOL`, substitute in all cases

File: src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java [hunk 8]
```diff
@@ -277,8 +281,8 @@ public void testExistingTargetFileXmlOutput() throws Exception {
     public void testExistingTargetFilePlainOutput() throws Exception {
 
         exit.checkAssertionAfterwards(() -> {
-            assertEquals(String.format(Locale.ROOT, "Starting audit...%n"
-                    + "Audit done.%n"), systemOut.getLog());
+            assertEquals(String.format(Locale.ROOT, auditStartMessage.getMessage() + "%n"
+                    + auditFinishMessage.getMessage() + "%n"), systemOut.getLog());
```
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java]: do we need `String.format(Locale.ROOT` , please review other cases too.

File: src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java (hunk 6 above)
Comment (subkrish) [inline src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java]: Done.

File: src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java (hunk 7 above)
Comment (subkrish) [inline src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java]: Done.

File: src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java (hunk 8 above)
Comment (subkrish) [inline src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java]: Done.

File: src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java [hunk 9]
```diff
@@ -468,24 +487,25 @@ public void testLoadPropertiesIoException() throws Exception {
...
         }
     }
 
     @Test
     public void testCreateListenerIllegalStateException() throws Exception {
         final Method method = Main.class.getDeclaredMethod("createListener", String.class,
             String.class);
+        final LocalizedMessage loadPropertiesMessage = new LocalizedMessage(0,
+                Definitions.CHECKSTYLE_BUNDLE, "Main.createListener",
+                new String[] {"myformat", "plain", "xml"}, null, getClass(), null);
```
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java]: please move as close to usage as possible.

File: src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java [hunk 10]
```diff
@@ -495,13 +515,16 @@ public void testCreateListenerWithLocationIllegalStateException() throws Excepti
             String.class);
         method.setAccessible(true);
         final String outDir = "myfolder123";
+        final LocalizedMessage createListenerMessage = new LocalizedMessage(0,
+                Definitions.CHECKSTYLE_BUNDLE, "Main.createListener",
+                new String[] {"myformat", "plain", "xml"}, null, getClass(), null);
```
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java]: please move as close to usage as possible.

File: src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java (hunk 9 above)
Comment (subkrish) [inline src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java]: Done.

File: src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java (hunk 10 above)
Comment (subkrish) [inline src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java]: Done.

**Comment by romani** [comment]
@subkrish , thanks a lot for update.

**Pull request merged**